<a href="https://colab.research.google.com/github/Andersonek/Artificial-Inteligence/blob/main/Machine_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

import numpy as np
import pandas as pd

In [ ]:
DF=pd.read_csv('ENB2012_data.csv',delimiter=';')
DF2=pd.read_csv('ENB2012_data.csv',delimiter=';')
DF

Relative Compactness  Surface Area  ...  Cooling Load  House class
0                    0.98         514.5  ...         21.33            0
1                    0.98         514.5  ...         21.33            0
2                    0.98         514.5  ...         21.33            0
3                    0.98         514.5  ...         21.33            0
4                    0.90         563.5  ...         28.28            1
..                    ...           ...  ...           ...          ...
763                  0.64         784.0  ...         21.40            0
764                  0.62         808.5  ...         16.88            0
765                  0.62         808.5  ...         17.11            0
766                  0.62         808.5  ...         16.61            0
767                  0.62         808.5  ...         16.03            0

[768 rows x 11 columns]

In [ ]:
DF.sort_values('Roof Area',ascending=False)
New_DF=DF[["Relative Compactness","Roof Area"]]#COPY

In [ ]:
DF.loc[DF['Surface Area']>750]

Relative Compactness  Surface Area  ...  Cooling Load  House class
36                   0.66         759.5  ...         12.40            0
37                   0.66         759.5  ...         12.23            0
38                   0.66         759.5  ...         12.40            0
39                   0.66         759.5  ...         12.14            0
40                   0.64         784.0  ...         16.78            0
..                    ...           ...  ...           ...          ...
763                  0.64         784.0  ...         21.40            0
764                  0.62         808.5  ...         16.88            0
765                  0.62         808.5  ...         17.11            0
766                  0.62         808.5  ...         16.61            0
767                  0.62         808.5  ...         16.03            0

[192 rows x 11 columns]

In [ ]:
DF.dtypes

Relative Compactness         float64
Surface Area                 float64
Wall Area                    float64
Roof Area                    float64
Overall Height               float64
Orientation                    int64
Glazing Area                 float64
Glazing Area Distribution      int64
Heating Load                 float64
Cooling Load                 float64
House class                    int64
dtype: object

#Algorytm odległości euklidesowej


In [ ]:
DF2['Odleglosc Euklidesa']=0
DF2

Relative Compactness  Surface Area  ...  House class  Odleglosc Euklidesa
0                    0.98         514.5  ...            0                    0
1                    0.98         514.5  ...            0                    0
2                    0.98         514.5  ...            0                    0
3                    0.98         514.5  ...            0                    0
4                    0.90         563.5  ...            1                    0
..                    ...           ...  ...          ...                  ...
763                  0.64         784.0  ...            0                    0
764                  0.62         808.5  ...            0                    0
765                  0.62         808.5  ...            0                    0
766                  0.62         808.5  ...            0                    0
767                  0.62         808.5  ...            0                    0

[768 rows x 12 columns]

In [ ]:
DF.iloc[1][0]

0.98

In [ ]:
lenght=len(DF)
ind=round(0.75*len(DF))
tests=lenght-ind
Dev_DF=DF2[:ind]
Test_DF=DF2[ind:]
Dev=Dev_DF
Test=Test_DF[['Relative Compactness','Surface Area','Wall Area','Roof Area','Overall Height','Orientation','Glazing Area','Glazing Area Distribution','Heating Load', 'Cooling Load']]
Test['House class']=None
Test_DF.iloc[0][-4]
Test.iloc[0][-3]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


32.82

In [ ]:
n=int(input("Podaj liczbę wierszy do porównania: "))

Podaj liczbę wieszy do porównania: 7


In [ ]:
for line in range(tests):
  Odleglosc=[0]*ind
  for row in range(ind):
    p2=Dev["Cooling Load"].iloc[row]
    p3=Dev["Heating Load"].iloc[row]
    Odleglosc[row]=round(np.sqrt((Test.iloc[line][-3]-p2)**2+(Test.iloc[line][-2]-p3)**2),2)
    Dev['Odleglosc Euklidesa'][row]=Odleglosc[row]
  new=Dev.nsmallest(n,'Odleglosc Euklidesa')
  meta=new['House class'].value_counts()[:1].index.tolist()
  a=meta[0]
  Test['House class'].iloc[line]=a


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2882: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(ind

In [ ]:
Test

Relative Compactness  Surface Area  ...  Cooling Load  House class
576                  0.98         514.5  ...         32.96            1
577                  0.98         514.5  ...         33.13            1
578                  0.98         514.5  ...         33.94            1
579                  0.98         514.5  ...         33.78            1
580                  0.90         563.5  ...         38.35            1
..                    ...           ...  ...           ...          ...
763                  0.64         784.0  ...         21.40            0
764                  0.62         808.5  ...         16.88            0
765                  0.62         808.5  ...         17.11            0
766                  0.62         808.5  ...         16.61            0
767                  0.62         808.5  ...         16.03            0

[192 rows x 11 columns]

In [ ]:
Test['House class'].iloc[0]

1

In [ ]:
TP=0
TN=0
FP=0
FN=0
for rows in range(tests):
  if ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==1)):
    TP+=1
  elif ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==0)):
    FP+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==0)):
    TN+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==1)):
    FN+=1

print("TP: ",TP, "TN: ",TN, "FP: ",FP, "FN: ",FN, "Precyzja: ",round(TP/(TP+FP),2),"Recall: ", round(TP/(TP+FN),2),"F1: ", round(2*((TP/(TP+FP)*(TP/(TP+FN)))/((TP/(TP+FP))+(TP/(TP+FN)))),2),"Accuracy ",round((TP+TN)/tests,2))

TP:  100 TN:  87 FP:  0 FN:  5 Precyzja:  1.0 Recall:  0.95 F1:  0.98 Accuracy  0.97


In [ ]:
  Odleglosc=[0]*ind
  for row in range(ind):
    p2=Dev['Cooling Load'].iloc[row]
    p3=Dev['Heating Load'].iloc[row]
    #Odleglosc[row]=round(np.sqrt((Test.iloc[20][-3]-p2)**2+(Test.iloc[20][-2]-p3)**2),2)
    Dev['Odleglosc Euklidesa'][row]=round(np.sqrt((Test.iloc[20][-3]-p2)**2+(Test.iloc[20][-2]-p3)**2),2)
  new=Dev.nsmallest(n,'Odleglosc Euklidesa')
  meta=new['House class'].value_counts()[:1].index.tolist()
  a=meta[0]
  Test['House class'].iloc[5]=a
new[['House class','Odleglosc Euklidesa']]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2882: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, valu

House class  Odleglosc Euklidesa
549            1                    0
548            1                    1
550            1                    1
551            1                    1
112            1                    3
258            1                    3
304            1                    3

#Klasyfikator Bayes'a

In [ ]:
lenght=len(DF)
ind=round(0.75*len(DF))
tests=lenght-ind
Dev_DF=DF[:ind]
Test_DF=DF[ind:]
Dev=Dev_DF
Test=Test_DF[['Relative Compactness','Surface Area','Wall Area','Roof Area','Overall Height','Orientation','Glazing Area','Glazing Area Distribution','Heating Load', 'Cooling Load']]
Test['House class']=None
Test_DF.iloc[0][-4]
Test.iloc[0][-3]

32.82

In [ ]:
T5=0
T10=0
T15=0
T20=0
T25=0
T30=0
T35=0
T40=0
for rows in range(ind):
  a=Dev['Cooling Load'].iloc[rows]
  if (Dev['House class'].iloc[rows]==1):
    if (a<=5):
      T5=+1
    elif (a>5 and a<=10):
      T10+=1
    elif (a>10 and a<=15):
      T15+=1
    elif (a>15 and a<=20):
      T20+=1
    elif (a>20 and a<=25):
      T25+=1
    elif (a>25 and a<=30):
      T30+=1
    elif (a>30 and a<=35):
      T35+=1
    elif (a>35):
      T40+=1
print(T5,"\n",T10,"\n",T15,"\n",T20,"\n",T25,"\n",T30,"\n",T35,"\n",T40)
Dev['House class'].sum()

0 
 0 
 0 
 0 
 14 
 105 
 88 
 79


286

In [ ]:
P5=0
P10=0
P15=0
P20=0
P25=0
P30=0
P35=0
P40=0
for rows in range(ind):
  a=Dev['Heating Load'].iloc[rows]
  if (Dev['House class'].iloc[rows]==1):
    if (a<=5):
      P5=+1
    elif (a>5 and a<=10):
      P10+=1
    elif (a>10 and a<=15):
      P15+=1
    elif (a>15 and a<=20):
      P20+=1
    elif (a>20 and a<=25):
      P25+=1
    elif (a>25 and a<=30):
      P30+=1
    elif (a>30 and a<=35):
      P35+=1
    elif (a>35):
      P40+=1

print(P5,"\n",P10,"\n",P15,"\n",P20,"\n",P25,"\n",P30,"\n",P35,"\n",P40)
Dev['House class'].sum()

0 
 0 
 0 
 11 
 51 
 103 
 49 
 72


286

In [ ]:
F5=0
F10=0
F15=0
F20=0
F25=0
F30=0
F35=0
F40=0
for rows in range(ind):
  a=Dev['Cooling Load'].iloc[rows]
  if (Dev['House class'].iloc[rows]==0):
    if (a<=5):
      F5=+1
    elif (a>5 and a<=10):
      F10+=1
    elif (a>10 and a<=15):
      F15+=1
    elif (a>15 and a<=20):
      F20+=1
    elif (a>20 and a<=25):
      F25+=1
    elif (a>25 and a<=30):
      F30+=1
    elif (a>30 and a<=35):
      F35+=1
    elif (a>35):
      F40+=1

print(F5,"\n",F10,"\n",F15,"\n",F20,"\n",F25,"\n",F30,"\n",F35,"\n",F40)
ind-Dev['House class'].sum()

0 
 0 
 149 
 119 
 22 
 0 
 0 
 0


290

In [ ]:
N5=0
N10=0
N15=0
N20=0
N25=0
N30=0
N35=0
N40=0
for rows in range(ind):
  a=Dev['Heating Load'].iloc[rows]
  if (Dev['House class'].iloc[rows]==0):
    if (a<=5):
      N5=+1
    elif (a>5 and a<=10):
      N10+=1
    elif (a>10 and a<=15):
      N15+=1
    elif (a>15 and a<=20):
      N20+=1
    elif (a>20 and a<=25):
      N25+=1
    elif (a>25 and a<=30):
      N30+=1
    elif (a>30 and a<=35):
      N35+=1
    elif (a>35):
      N40+=1

print(N5,"\n",N10,"\n",N15,"\n",N20,"\n",N25,"\n",N30,"\n",N35,"\n",N40)
ind-Dev['House class'].sum()

0 
 20 
 212 
 58 
 0 
 0 
 0 
 0


290

In [ ]:
for i in range(tests):
  a=Test["Cooling Load"].iloc[i]
  b=Test['Heating Load'].iloc[i]
  if (a<=5):
    Yes=T5
    No=F5
  elif (a>5 and a<=10):
    Yes=T10
    No=F10
  elif (a>10 and a<=15):
    Yes=T15
    No=F15
  elif (a>15 and a<=20):
    Yes=T20
    No=F20
  elif (a>20 and a<=25):
    Yes=T25
    No=F25
  elif (a>25 and a<=30):
    Yes=T30
    No=F30
  elif (a>30 and a<=35):
    Yes=T35
    No=F35
  elif (a>35):
    Yes=T40
    No=F40
  if (b<=5):
    Tr=P5
    Fl=N5
  elif (b>5 and b<=10):
    Tr=P10
    Fl=N10
  elif (b>10 and b<=15):
    Tr=P15
    Fl=N15
  elif (b>15 and b<=20):
    Tr=P20
    Fl=N20
  elif (b>20 and b<=25):
    Tr=P25
    Fl=N25
  elif (b>25 and b<=30):
    Tr=P30
    Fl=N30
  elif (b>30 and b<=35):
    Tr=P35
    Fl=N35
  elif (b>35):
    Tr=P40
    Fl=N40
  Py=(Yes/Dev['House class'].sum())*(Tr/Dev['House class'].sum())*(Dev['House class'].sum()/ind)
  Pn=(No/(ind-Dev['House class'].sum()))*(Fl/(ind-Dev['House class'].sum()))*((ind-Dev['House class'].sum())/ind)
  PPY=Py/(Py+Pn)
  PPn=Pn/(Py+Pn)
  if PPY<PPn:
    Test['House class'].iloc[i]=0
  else:
    Test['House class'].iloc[i]=1

Test

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


Relative Compactness  Surface Area  ...  Cooling Load  House class
576                  0.98         514.5  ...         32.96            1
577                  0.98         514.5  ...         33.13            1
578                  0.98         514.5  ...         33.94            1
579                  0.98         514.5  ...         33.78            1
580                  0.90         563.5  ...         38.35            1
..                    ...           ...  ...           ...          ...
763                  0.64         784.0  ...         21.40            0
764                  0.62         808.5  ...         16.88            0
765                  0.62         808.5  ...         17.11            0
766                  0.62         808.5  ...         16.61            0
767                  0.62         808.5  ...         16.03            0

[192 rows x 11 columns]

In [ ]:
TP=0
TN=0
FP=0
FN=0
for rows in range(tests):
  if ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==1)):
    TP+=1
  elif ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==0)):
    FP+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==0)):
    TN+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==1)):
    FN+=1

print("TP: ",TP, "TN: ",TN, "FP: ",FP, "FN: ",FN, "Precyzja: ",round(TP/(TP+FP),2),"Recall: ", round(TP/(TP+FN),2),"F1: ", round(2*((TP/(TP+FP)*(TP/(TP+FN)))/((TP/(TP+FP))+(TP/(TP+FN)))),2),"Accuracy ",round((TP+TN)/tests,2))

TP:  96 TN:  87 FP:  0 FN:  9 Precyzja:  1.0 Recall:  0.91 F1:  0.96 Accuracy  0.95


Widać nieco mniejszą precyzję tej metody. Zapewne jest to spowodowane dobrze dobranym przedziałem (krok co 5). Jednak wciąż błąd pozostaje niewielki (zaledwie 5%)

#Regresja liniowa


In [ ]:
lenght=len(DF)
ind=round(0.75*len(DF))
tests=lenght-ind
Dev_DF=DF[:ind]
Test_DF=DF[ind:]
Dev=Dev_DF
Test=Test_DF[['Relative Compactness','Surface Area','Wall Area','Roof Area','Overall Height','Orientation','Glazing Area','Glazing Area Distribution','Heating Load', 'Cooling Load']]
Test['House class']=None
Test_DF.iloc[0][-4]
Test.iloc[0][-3]

32.82

In [ ]:
def sigmod(x):
  return 1/(1+np.exp(-x))

def logistic_loss(y, y_hat):
  return -np.mean(y*np.log(y_hat)+(1-y)*np.log(1-y_hat))

x_train=Dev[["Heating Load","Cooling Load"]]
x_test=Test[['Heating Load','Cooling Load']]
y_train=Dev['House class']
y_test=Test['House class']

y_train=y_train.values.reshape(-1,1)
y_test=y_test.values.reshape(-1,1)


W=np.zeros((2,1))
b=np.zeros((1,1))

learning_rate = 0.01

n=len(y_train)

for epoch in range(1000):
  Z= np.dot(x_train,W)+b
  A=sigmod(Z)
  loss = logistic_loss(y_train,A)
  dz= A - y_train

  dw= 1/n*np.dot(x_train.T,dz)

  db = np.sum(dz)

  W=W-learning_rate*dw
  b=b-learning_rate*db
  if (epoch%100==0):
    print(loss)

0.6931471805599453
0.05633711168876846
0.03941684309866658
0.032443199392472345
0.028451717784261914
0.025793033817777766
0.02385850286643374
0.02236688294365531
0.02116873129703607
0.020176621010978685


In [ ]:
for i in range(tests):
  if (sigmod(np.dot(x_test.iloc[i], W) + b)>0.5):
    Test['House class'].iloc[i]=1
  else:
    Test['House class'].iloc[i]=0

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [ ]:
x_test

Heating Load  Cooling Load
576         32.82         32.96
577         32.71         33.13
578         32.24         33.94
579         32.72         33.78
580         35.84         38.35
..            ...           ...
763         17.88         21.40
764         16.54         16.88
765         16.44         17.11
766         16.48         16.61
767         16.64         16.03

[192 rows x 2 columns]

In [ ]:
sigmod(np.dot(x_test.iloc[1], W) + b)

array([[0.99999278]])

In [ ]:
TP=0
TN=0
FP=0
FN=0
for rows in range(tests):
  if ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==1)):
    TP+=1
  elif ((Test['House class'].iloc[rows]==1) and (Test_DF['House class'].iloc[rows]==0)):
    FP+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==0)):
    TN+=1
  elif ((Test['House class'].iloc[rows]==0) and (Test_DF['House class'].iloc[rows]==1)):
    FN+=1

print("TP: ",TP, "TN: ",TN, "FP: ",FP, "FN: ",FN, "Precyzja: ",round(TP/(TP+FP),2),"Recall: ", round(TP/(TP+FN),2),"F1: ", round(2*((TP/(TP+FP)*(TP/(TP+FN)))/((TP/(TP+FP))+(TP/(TP+FN)))),2),"Accuracy ",round((TP+TN)/tests,2))

TP:  98 TN:  87 FP:  0 FN:  7 Precyzja:  1.0 Recall:  0.93 F1:  0.97 Accuracy  0.96


#Podsumowanie

Najlepszym modelem okazał się model wykorzystujący odległość euklidesa, poprawność tutaj wynosiła 97%. Drugim z koleji był model regresji logistycznej. Jego dokładność wynosiła 96%. Ostatnim okazał się model klasyfikatora prawdopodobieństwa z wynikiem 95%. Jak widać wszystkie modele okazały się bardzo dokładne, różnice wynosiły zaledwie 1pkt p.